Succeed even if the trailing slash is omitted

esoergel opened the pull request:
## Summary
https://dimagi-dev.atlassian.net/browse/QA-3148

I'd like to be able to do this with a better URL pattern, but don't think it's possible.  This is at least explicit, the duplication is minimal (it's essentially registering another valid URL pattern), and I think it's worth the trade-off to provide a better user experience.

## Feature Flag
"Enable the v0.6 Case API"

## Product Description
Succeed even if the trailing slash is omitted

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
This is in response to a QA ticket

### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
This feature is unreleased and in QA

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

esoergel committed:
Succeed even if the trailing slash is omitted

kaapstorm commented on corehq/apps/api/urls.py:
```diff
@@ -119,7 +120,8 @@ def api_url_patterns():
     yield url(r'v0.5/odata/forms/(?P<config_id>[\w\-:]+)/\$metadata$',
               ODataFormMetadataView.as_view(), name=ODataFormMetadataView.urlname)
-    yield url(r'v0.6/case/', include(case_api_urlpatterns))
+    yield url(r'v0.6/case$', case_api)  # We don't want Django to redirect to add a trailing slash
+    yield url(r'v0.6/case/?', include(case_api_urlpatterns))
```
This should work. It reduces duplication and allows you to drop `case_api_urlpatterns`, but it's not as simple.
```suggestion
    yield url(r'v0.6/case(?:/(?P<case_id>[\w-]+)?/?)?$', case_api, name='case_api')
```

millerdev replied:
Alternate that does not match `v0.6/case//`:
```suggestion
    yield url(r'v0\.6/case(?:/(?P<case_id>[\w-]+))?/?$', case_api, name='case_api')
```

kaapstorm replied:
Ah! Nice.

esoergel replied:
Thanks you two, just tried it out (including with tests) and it works like a charm.  The `\.` didn't escape my notice either - nice touch.

esoergel committed:
Use fancy URL pattern to handle trailing slashes

With PUT and POST requests, omitting the trailing slash causes a
redirect, which clients typically don't handle well (it strips the POST
data).  Shout out to Norman and Daniel for the syntax.
